### PR TITLE
[#PAB-342] implement the new data model changes

### PR DESCRIPTION
**Pull request by cyrusdobbs**
https://trello.com/c/HIYrpcys/342-goal-data-model-changes-implementation

### Change description
Implemented the new data model changes to the entities. 

- [x] Unit tests and other appropriate tests added or updated
- [NA] README and other documentation has been updated / added (if needed)
- [X] Commit messages are meaningful and follow good commit message guidelines (e.g. "FS-XXXX: Add margin to nav items preventing overlapping of logo")

### How to test
_If manual testing is needed, give suggested testing steps_

### CODE REVIEW

**Comment by cyrusdobbs** [comment]
First draft done

TODO:

- Add submission table
- Implement `populate_db` changes
- Test ingest of the data model spreadsheet

**Review by nsnape — CHANGES_REQUESTED** [review]

File: core/db/entities.py [hunk 1]
```diff
@@ -71,269 +69,236 @@ class Package(BaseModel):
...
+    # TODO: This mapping doesn't exist now
+    # package_id: Mapped[int] = sqla.orm.mapped_column(sqla.ForeignKey("package_dim.id"), nullable=True)
 
-    # Unique index for data integrity. There can't be multiple direct fund rows for a single project with
-    # the same date range and direct fund metrics.
+    primary_intervention_theme = sqla.Column(sqla.String(), nullable=False)
+    location_multiplicity = sqla.Column(
+        sqla.Enum(const.MultiplicityEnum, name="project_location_multiplicity"), nullable=False
+    )
+    # TODO: configured as an array of strings. is this ok? can we / should we validate for this?
```
Comment (nsnape) [inline core/db/entities.py:171]: Yeah let's leave it in and try it out

File: core/db/entities.py [hunk 2]
```diff
@@ -71,269 +69,236 @@ class Package(BaseModel):
...
+
+    progress_records: Mapped[List["ProjectProgress"]] = sqla.orm.relationship()
+    funding: Mapped[List["Funding"]] = sqla.orm.relationship()
+    funding_comments: Mapped[List["FundingComment"]] = sqla.orm.relationship()
+    private_investments: Mapped[List["PrivateInvestment"]] = sqla.orm.relationship()
+    outputs: Mapped[List["OutputData"]] = sqla.orm.relationship()
+    outcomes: Mapped[List["OutcomeData"]] = sqla.orm.relationship()
+    risks: Mapped[List["RiskRegister"]] = sqla.orm.relationship()
+
+    # TODO: does this unique index look right?
```
Comment (nsnape) [inline core/db/entities.py:182]: Yeah I think this will be fine

File: core/db/entities.py [hunk 3]
```diff
@@ -446,7 +417,11 @@ class RiskRegister(BaseModel):
 
     __tablename__ = "risk_register"
 
-    project_id: Mapped[int] = sqla.orm.mapped_column(sqla.ForeignKey("project_dim.id"), nullable=False)
+    submission_id: Mapped[int] = sqla.orm.mapped_column(sqla.ForeignKey("submission_dim.id"), nullable=False)
+    # TODO: add logic to support nullable but one of programme or project id requred
```
[the remark this answers is not in the document]
Comment (nsnape) [reply core/db/entities.py:421]: in this case, that will be added to the __table_args__ tuple that we already use to define the unique index.

**Review by nsnape — CHANGES_REQUESTED** [review]
Great stuff, sorry about the inundation of comments, there will be a bit of an iterative review process for this I think :)

File: core/db/entities.py [hunk 4]
```diff
@@ -71,269 +69,236 @@ class Package(BaseModel):
...
+    # package_id: Mapped[int] = sqla.orm.mapped_column(sqla.ForeignKey("package_dim.id"), nullable=True)
 
-    # Unique index for data integrity. There can't be multiple direct fund rows for a single project with
-    # the same date range and direct fund metrics.
+    primary_intervention_theme = sqla.Column(sqla.String(), nullable=False)
+    location_multiplicity = sqla.Column(
+        sqla.Enum(const.MultiplicityEnum, name="project_location_multiplicity"), nullable=False
+    )
+    # TODO: configured as an array of strings. is this ok? can we / should we validate for this?
+    locations = sqla.Column(sqla.ARRAY(sqla.String), nullable=False)
```
Comment (nsnape) [inline core/db/entities.py:172]: also, can we please add a field to capture the "Are you providing a GIS map (see guidance) with your return?" column. Should be nullable and a yes/no enum.

**Review by nsnape — CHANGES_REQUESTED** [review]

File: core/db/entities.py [hunk 5]
```diff
@@ -71,269 +69,236 @@ class Package(BaseModel):
...
+    funding_source_name = sqla.Column(sqla.String(), nullable=False)
+    funding_source_type = sqla.Column(sqla.String(), nullable=False)
+    # TODO: nullable as per dm ss
+    secured = sqla.Column(sqla.Enum(const.YesNoEnum, name="funding_secured"), nullable=False)
+    # TODO: nullable as per dm ss
+    spend_before_reporting_commenced = sqla.Column(sqla.Float(), nullable=True)
+    # TODO: Should this be a lookup ?
+    reporting_period = sqla.Column(sqla.String(), nullable=True)
+    # TODO: nullable as per dm ss
+    spend_for_reporting_period = sqla.Column(sqla.Float(), nullable=True)
```
[the remark this answers is not in the document]
Comment (cyrusdobbs) [reply core/db/entities.py:243]: I've also changed spend_before_reporting_commenced to nullable with default 0

[the remark this answers is not in the document]
Comment (cyrusdobbs) [reply core/db/entities.py:243]: and spend_beyond_fund_lifetime

File: core/db/entities.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nsnape) [reply core/db/entities.py]: oh, no actually it needs to be XOR, one OR the other but not both

**Review by nsnape — CHANGES_REQUESTED** [review]

**Review by nsnape — CHANGES_REQUESTED** [review]

File: core/db/entities.py [hunk 6]
```diff
@@ -344,30 +296,26 @@ class OutputData(BaseModel):
 
     __tablename__ = "output_data"
 
+    submission_id: Mapped[int] = sqla.orm.mapped_column(sqla.ForeignKey("submission_dim.id"), nullable=False)
     project_id: Mapped[int] = sqla.orm.mapped_column(sqla.ForeignKey("project_dim.id"), nullable=False)
+    output_id: Mapped[int] = sqla.orm.mapped_column(sqla.ForeignKey("output_dim.id"), nullable=False)
+
```
Comment (nsnape) [inline core/db/entities.py:302]: can you add "additional_information" nullable string field to OutputData?

File: core/db/entities.py [hunk 7]
```diff
@@ -397,33 +345,39 @@ class OutcomeData(BaseModel):
...
     start_date = sqla.Column(sqla.DateTime(), nullable=False)
     end_date = sqla.Column(sqla.DateTime(), nullable=False)
-
-    # TODO: as per comment on output
     unit_of_measurement = sqla.Column(sqla.String(), nullable=False)
     geography_indicator = sqla.Column(
         sqla.Enum(const.GeographyIndicatorEnum, name="outcome_data_geography"), nullable=False
     )
     amount = sqla.Column(sqla.Float(), nullable=False)
     state = sqla.Column(sqla.Enum(const.StateEnum, name="outcome_data_state"), nullable=False)
```
Comment (nsnape) [inline core/db/entities.py:361]: we also need to add a new field to this one: "higher_frequency", nullable string

**Review by nsnape — CHANGES_REQUESTED** [review]

File: core/db/entities.py [hunk 8]
```diff
@@ -31,309 +34,258 @@ def to_dict(self):
...
-    answer_3 = sqla.Column(sqla.String(), nullable=True)
-    answer_4 = sqla.Column(sqla.String(), nullable=True)
-    answer_5 = sqla.Column(sqla.String(), nullable=True)
-    answer_6 = sqla.Column(sqla.String(), nullable=True)
+    project_name = sqla.Column(sqla.String(), nullable=False)
+    primary_intervention_theme = sqla.Column(sqla.String(), nullable=False)
+    location_multiplicity = sqla.Column(
+        sqla.Enum(const.MultiplicityEnum, name="project_location_multiplicity"), nullable=False
+    )
+    # locations = sqla.Column(sqla.ARRAY(sqla.String), nullable=False)  # ARRAY not supported by SQLite
```
Comment (nsnape) [inline core/db/entities.py:169]: thinking about it, let's not use the array at all - that reply's far to much on users entering in some valid format. Just a string is fine (can remove the comment).

**Pull request merged**